test/jdk/java/net/vthread/BlockingSocketOps.java

```diff
@@ -191,7 +192,11 @@ void testSocketReadPeerClose2() throws Exception {
                 // read from s1 should block, then throw
                 try {
                     int n = s1.getInputStream().read();
-                    fail("read " + n);
+                    if (!Platform.isAix()) {
+                        fail("read " + n);
+                    } else {
+                        assertTrue(n == -1);
```

Comment:
This doesn't look right, read should not return -1 here.

Reply:
I believe we get [here](https://github.com/openjdk/jdk/blob/master/src/java.base/unix/native/libnio/ch/SocketDispatcher.c#L44) with EAGAIN, but not ECONNRESET. So the -1 indicates that the read has failed.

My feeling is that the defined behaviour is not totally clear. [From setSockOpt](https://linux.die.net/man/3/setsockopt) (emphasis added by me):

> SO_LINGER
>   Lingers on a close() _if data is present_.

This does not define what happens if no data is present. In my testing, AIX behaved exactly like it does in `testSocketReadPeerClose1` so my understanding is that SO_LINGER had essentially no effect because there is no data waiting to be sent. As I see it, this test reduces to `testSocketReadPeerClose1` on AIX, so the test should be the same. Another option would be to skip it entirely on AIX.

Reply:
There are several tests in both the java/net and java/nio/channels tree that setup the conditions for a "hard reset", e.g. SocketChannel/ConnectionReset.java. I'm curious if these tests also fail on AIX.

Reply:
Thanks for mentioning it. I'll take a look at those to see if they also fail.

Reply:
That test passes. I'll take a look into the differences between the two tests.

Reply:
The long standing spec for SO_LINGER is "Enabling the option with a timeout of zero does a forceful close immediately". The wording isn't quite right but it is trying to say that if the timeout is set to zero then calling the close method will cause a forceful close. There are several tests that use this so there might be other failures on AIX.

Reply:
The ConnectionReset test you mentioned was very helpful. Comparing the pass there with the failure in `testSocketReadPeerClose2` led me to a better fix. I believe the test failure actually had to do with the order that the close, and setSoLinger methods were being executed in.